lib/rdc.js

```diff
@@ -0,0 +1,97 @@
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+const rdc = (source, destination, options, callback) => {
+  if (typeof callback !== 'function') {
+    if (typeof options !== 'function') throw new Error('ERR_INVALID_CALLBACK');
+    callback = options;
+    options = {};
+  }
+
+  const depth = options.depth;
+  const Transform = options.change.transform;
+  const readStreamOptions = options.change.readStreamOptions;
+  const writeStreamOptions = options.change.writeStreamOptions;
+  const filter = options.filter;
+  const stop = options.stopOnError;
+  const errorLog = options.errorLog;
+  const dereference = options.dereference;
+
+  const error = err => {
+    if (stop) throw new Error(err);
+    else if (errorLog) errorLog.write(JSON.stringify(err));
+    else callback(err);
+  };
+
+  const copyDirectory = target => {
+    fs.mkdirSync(target, err => {
+      if (err) error(err);
+    });
+  };
+
+  const copyFile = (current, target) => {
+    if (Transform) {
+      fs.createReadStream(current, readStreamOptions)
+        .pipe(new Transform())
+        .pipe(fs.createWriteStream(target, writeStreamOptions));
+    } else {
+      fs.copyFile(current, target, err => {
+        if (err) error(err);
+      });
+    }
+  };
+
+  const copyLink = (current, target) => {
+    fs.readlink(current, (err, linkString) => {
+      if (err) error(err);
+      if (dereference) {
+        copyDirectory(target);
+        start(path.resolve(path.dirname(current), linkString), target);
+      } else {
+        fs.symlink(linkString, target, err => {
+          if (err) error(err);
+        });
+      }
+    });
+  };
+
+  const start = (curDir = source, dest = destination, curDepth = 1) => {
+    if (curDepth > depth) return;
+    fs.readdir(curDir, (err, items) => {
```

Comment:
I don't see where you call the callback when you succeeded in copying.

Reply:
Yes, it is my mistake. What do you think about using counter to check the last file coping and then call callback(null)? or think about another solution?

Reply:
That is related to this [comment](https://github.com/metarhia/common/pull/251/files#r256290296).
This is example of basic recursive directory copying.
```js
const copydir = (src, dest, cb) => {
  fs.readdir(dest, (err, files) => {
    if (err) {
      cb(err);
      return;
    }
    metasync.each(files, (file, cb) => {
      fs.stat(file, (err, stats) => {
        if (err) {
          cb(err);
          return;
        }
        const srcFile = path.join(src, file);
        const destFile = path.join(dest, file);
        if (stats.isDirectory()) {
          fs.mkdir(destFile, err => {
            if (err) cb(err);
            else copydir(srcFile, destFile, cb);
          });
        } else {
          fs.copyFile(srcFile, destFile, cb);
        }
      });
    }, cb);
  });
};
```